src/ripple/app/misc/impl/AMMUtils.cpp

```diff
@@ -287,17 +287,18 @@ initializeFeeAuctionVote(
     Issue const& lptIssue,
     std::uint16_t tfee)
 {
+    bool fixInnerObj(view.rules().enabled(fixInnerObjTemplate));
     // AMM creator gets the voting slot.
     STArray voteSlots;
-    STObject voteEntry{sfVoteEntry};
+    STObject voteEntry{sfVoteEntry, fixInnerObj};
     if (tfee != 0)
         voteEntry.setFieldU16(sfTradingFee, tfee);
     voteEntry.setFieldU32(sfVoteWeight, VOTE_WEIGHT_SCALE_FACTOR);
     voteEntry.setAccountID(sfAccount, account);
     voteSlots.push_back(voteEntry);
     ammSle->setFieldArray(sfVoteSlots, voteSlots);
     // AMM creator gets the auction slot for free.
-    auto& auctionSlot = ammSle->peekFieldObject(sfAuctionSlot);
+    STObject auctionSlot{sfAuctionSlot, fixInnerObj};
```

Comment:
Our current setup for amendments works well when modifying and add transactors. It works less well with bug fixes - especially bug fixes that effect lower level classes. We've run into this several times with STAmount and number classes, and now we're running into it with STObject. I dislike passing in a bool into the constructor like this patch does, although I understand why you did it.

What do you think about a `thread_local` variable that holds the current `Rules` (defaulting to a null ptr if there is no transaction active)? The variable would be setup on entry to a transaction and torn down on exit. The problems I see with this are:

1. It won't work with coroutines, since these can be resumed on different threads. I don't believe this is a problem now, but it could let bugs sneak in later if we use coroutines.
2. It wouldn't work with recursive transactions (which we don't have, but "batch" transactions may have a problem depending how they are implemented). This is fixable with a stack but could be another source of errors.
3. It may break leveling. This `thread_local` needs to sit at a lowest level of the hierarchy of a class that would use it. For now it could sit at the `protocol` level, which is fine. But later this could be an issue.

All in all I think I favor going in this directions, but thought I'd open it up for discussion. I'm leaving it as a comment on code because top-level comments cannot have threaded discussions.

Reply:
I like this idea. This could be made into a more general solution then. The overloaded `STObject` constructor with `bool` argument is not needed and existing `STObject(SField const&)` constructor can be used instead. This variable can set/torn down in `applyTransaction()` and some added AMM code can be removed. This is somewhat similar to `thread_local Number::rounding_mode Number::mode_`. Does this sound right?

If `applyTransaction()` is the right place then wouldn't it address 2?

Reply:
@seelabs Because, as far as I know, inner objects are always and only created in transactor classes and related helper functions, what do you think about undoing the `STObject` ctor changes, and adding a `makeInnerObject` function that has a `Rules` parameter?
```
STObject makeInnerObject(SField const& name, Rules const& rules)
{
    STObject obj{name};
    if (rules.enabled(fixInnerObjTemplate))
    {
        SOTemplate const* elements =
            InnerObjectFormats::getInstance().findSOTemplateBySField(name);
        if (elements)
            set(*elements);
    }
}
```
Both `STObject` and `Rules` live in `src/ripple/protocol`, so there are no levelization implications.

If a new amendment is added that uses inner objects, the `if` could check for all of them.

Reply:
Then each inner object has to be explicitly constructed this way rather then implicitly handled via `STObject` constructor. Why is this a better approach? To avoid 1-3 issues that @seelabs mentioned?

Reply:
Yeah, it avoids those issues, but it may not be a better solution. Maybe you could just pass `Rules` (or more specifically `std::optional<Rules>` into the current ctor and not add the helper).

Reply:
Actually, I would rather use the helper. The constructor is obscure.

Reply:
I like @ximinez idea of passing `Rules` - good thinking Ed.